An update to the install instructions that make use of tabbing. (#2155)
knative/docs: docs/install/Knative-with-any-k8s.md

---

> title: "Installing Knative"

Reviewer: Global: numbered lists should all be 1; automatic list numbering works on the site now.

linkTitle: "Installing Knative"
weight: 15
type: "docs"
---

This guide walks you through the installation of the latest version of Knative.

> Knative has several pieces, which can be installed and used independently or together.

Reviewer 2: ```suggestion
Knative has two components, which can be installed and used independently or together. Knative also has a monitoring bundle that can optionally be installed to make the components more observable.
```

Author: I tweaked this a bunch, but done.

To help you pick and choose the pieces that are right for you, here is a brief
description of each:
 - [**Serving**](#installing-the-serving-component)(_stable_) provides an abstraction for stateless request-based scale-to-zero services.
 - [**Eventing**](#installing-the-eventing-component)(_alpha_) provides abstractions to enable binding event sources (e.g. Github Webhooks, Kafka) and consumers (e.g. Kubernetes or Knative Services).
 - [**Monitoring**](#installing-the-monitoring-bundle)(_alpha_) provides a bundle of standard tooling that can be used to get visibility into the health of the software running on Knative.

## Before you begin

This guide assumes that you want to install an upstream Knative release on a Kubernetes cluster. A growing number of vendors are offering managed Knative services, see [here](./vendor-options.md) for a list of the known offerings.

> Knative requires a Kubernetes cluster v1.15 or newer, as well as a compatible

Reviewer 2: ```suggestion
Knative < version > requires a Kubernetes cluster v1.15 or newer, as well as a compatible
```

It would also be nice to have a consolidated < min_k8s_version > variable.

`kubectl`. This guide assumes that you've already created a Kubernetes cluster,
and that you are using bash in a Mac or Linux environment; some commands will
need to be adjusted for use in a Windows environment.

<!-- TODO: Link to provisioning guide for advanced installation -->

## Installing the Serving component

The following commands install the Knative Serving component.

1. Install the Custom Resource Definitions (aka CRDs):

   ```bash
   kubectl apply --filename https://github.com/knative/serving/releases/download/{{< version >}}/serving-crds.yaml
   ```

1. Install the core components of Serving (see below for optional extensions):

   ```bash
   kubectl apply --filename https://github.com/knative/serving/releases/download/{{< version >}}/serving-core.yaml
   ```

> 

Reviewer 2: ```suggestion
  <!-- TODO: Link to document/diagram describing what is a networking layer.
```

      --namespace knative-serving \
      --type merge \
      --patch '{"data":{"ingress.class":"contour.ingress.networking.knative.dev"}}'
    ```

3. Fetch the External IP or CNAME

   ```bash
   kubectl --namespace projectcontour get service envoy-external
   ```

   Save this for configuring DNS below.

{{< /tab >}}

{{% tab name="Gloo" %}}
_For a detailed guide on Gloo integration, see [here](https://docs.solo.io/gloo/latest/installation/knative/)_

The following commands install Gloo and enables its Knative integration:

1. Make sure `glooctl` is installed (version 1.3.x and higher recommended):

   ```bash
   glooctl version
   ```

   If it is not installed, follow the instructions [here](https://docs.solo.io/gloo/latest/installation/knative/#install-command-line-tool-cli)

2. Install Gloo and the Knative integration:

    ```bash
    glooctl install knative --install-knative=false
    ```

3. Fetch the External IP or CNAME

   ```bash
   kubectl get svc -ngloo-system knative-external-proxy
   ```

   Save this for configuring DNS below.

{{< /tab >}}

{{% tab name="Istio" %}}
The following commands install Istio and enables its Knative integration:

> 1. [Installing Istio for Knative](./installing-istio.md)

Reviewer 2: All of the other guides have commands inline. Linking out to a different document within a tab and expecting the user to come back to the same tab feels like a bad experience. 

Can we have a basic install command and have a `_For a detailed guide on Istio integration, see [here]` like we do for Gloo?

Author: This isn't that different from what we have today, so I'd be inclined to leave this as-is and iterate with further incremental improvements.  I'll open an issue and `TODO(#1234)` it.

Reviewer: We could also go the route of including the content from the installing istio page here using readfile.

Reviewer 2: The istio page is too large for this section. We would need to slim it down a bit.

Reviewer 2: I am okay with the issue that was created.

The following commands install Kourier and enables its Knative integration:

1. Install the Knative Kourier controller:

   ```bash
   kubectl apply --filename https://raw.githubusercontent.com/knative/serving/{{< version >}}/third_party/kourier-latest/kourier.yaml
   ```

2. To configure Knative Serving to use Kourier by default:

    ```bash
    kubectl patch configmap/config-network \
      --namespace knative-serving \
      --type merge \
      --patch '{"data":{"ingress.class":"kourier.ingress.networking.knative.dev"}}'
    ```

3. Fetch the External IP or CNAME

   ```bash
   kubectl --namespace kourier-system get service kourier
   ```

   Save this for configuring DNS below.

{{< /tab >}}
{{< /tabs >}}

1. Configure DNS

   <!-- This indentation is important for things to render properly. -->
   {{< tabs name="serving_dns" >}}
{{% tab name="Magic DNS (xip.io)" %}}
We ship a simple Kubernetes Job called "default domain" that will (see caveats) configure Knative Serving to use <a href="http://xip.io">xip.io</a> as the default DNS suffix.

```bash
kubectl apply --filename https://github.com/knative/serving/releases/download/{{< version >}}/serving-default-domain.yaml
```

> **Caveat**: This will only work if the cluster LoadBalancer service exposes an IPv4 address, so it will not work with IPv6 clusters, AWS, or local setups like Minikube.  For these, see "Using a real DNS record" (below).

Reviewer 2: ```suggestion
**Caveat**: This will only work if the cluster LoadBalancer service exposes an IPv4 address, so it will not work with IPv6 clusters, AWS, or local setups like Minikube.  For these, see "Using a real DNS record".
```

I don't think this is (below) when it is in tab form?

Author: Good catch :D

{{< /tab >}}

{{% tab name="TLS wildcard support" %}}

If you plan to use cert-manager's support for provisioning wildcard certificates, then the most efficient way to provision certificates is with the namespace wildcard certificate controller.  The following command will install the components needed to provision wildcard certificates in each namespace:

```bash
kubectl apply --filename https://github.com/knative/serving/releases/download/{{< version >}}/serving-nscert.yaml
```

{{< /tab >}}
{{< /tabs >}}

### Getting started with Serving

Deploy your first app with the [getting started with Knative app deployment](../serving/getting-started-knative-app.md) guide.  You can also find a number of samples for Knative Serving [here](../eventing/samples/).

## Installing the Eventing component

The following commands install the Knative Eventing component.

1. Install the Custom Resource Definitions (aka CRDs):

   ```bash
   kubectl apply --filename https://github.com/knative/eventing/releases/download/{{< version >}}/eventing-crds.yaml
   ```

1. Install the core components of Eventing (see below for optional extensions):

   ```bash

>    kubectl apply --filename https://github.com/knative/eventing/releases/download/{{< version >}}/eventing-core.yaml

Author: (context from me) I want to move away from installing bundles, in favor of the tabbed approach.  I doubt folks will want the in memory channel on production clusters, and we don't document how to not have that.  I hope @vaikas is aligned with that sentiment?

>    kubectl apply --filename https://github.com/knative/eventing/releases/download/{{< version >}}/eventing-core.yaml

Reviewer: ah, okay. I only knew what we had been telling folks to install previously.

   ```

   Save this for configuring DNS below.

{{< /tab >}}

{{% tab name="Contour" %}}
The following commands install Contour and enable its Knative integration.

1. Install the Knative Contour controller (includes Contour):

   ```bash
   kubectl apply --filename https://raw.githubusercontent.com/knative/serving/{{< version >}}/third_party/contour-latest/contour.yaml
   ```

1. To configure Knative Serving to use Contour by default:

    ```bash
    kubectl patch configmap/config-network \
      --namespace knative-serving \
      --type merge \
      --patch '{"data":{"ingress.class":"contour.ingress.networking.knative.dev"}}'
    ```

1. Fetch the External IP or CNAME:

   ```bash
   kubectl --namespace projectcontour get service envoy-external
   ```

   Save this for configuring DNS below.

{{< /tab >}}

{{% tab name="Gloo" %}}
_For a detailed guide on Gloo integration, see [Installing Gloo for Knative](https://docs.solo.io/gloo/latest/installation/knative/) in the Gloo documentation._

The following commands install Gloo and enable its Knative integration.

1. Make sure `glooctl` is installed (version 1.3.x and higher recommended):

   ```bash
   glooctl version
   ```

>    If it is not installed, follow the instructions [here](https://docs.solo.io/gloo/latest/installation/knative/#install-command-line-tool-cli)

Reviewer 3: can we please add the following instruction here:

```bash
curl -sL https://run.solo.io/gloo/install | sh
export PATH=$HOME/.gloo/bin:$PATH
```

Reviewer 3: this is for installing glooctl

1. Install Gloo and the Knative integration:

    ```bash
    glooctl install knative --install-knative=false
    ```

1. Fetch the External IP or CNAME:

   ```bash

>    kubectl get svc -ngloo-system knative-external-proxy

Reviewer 3: we can also provide `glooctl proxy url --name knative-external-proxy` as an alternative here

---
title: "Installing Knative"
linkTitle: "Installing Knative"
weight: 15
type: "docs"
---

This guide walks you through the installation of the latest version of Knative.

Knative has two components, which can be installed and used independently or together.
To help you pick and choose the pieces that are right for you, here is a brief
description of each:
 - [**Serving**](#installing-the-serving-component)(_stable_) provides an abstraction for stateless request-based scale-to-zero services.
 - [**Eventing**](#installing-the-eventing-component)(_alpha_) provides abstractions to enable binding event sources (e.g. Github Webhooks, Kafka) and consumers (e.g. Kubernetes or Knative Services).

Knative also has a [**Monitoring bundle**](#installing-the-monitoring-bundle)(_alpha_) which provides standard tooling that can be used to get visibility into the health of the software running on Knative.

## Before you begin

> This guide assumes that you want to install an upstream Knative release on a Kubernetes cluster. A growing number of vendors are offering managed Knative services, see [here](./vendor-options.md) for a list of the known offerings.

Reviewer: ```suggestion
This guide assumes that you want to install an upstream Knative release on a Kubernetes cluster. A growing number of vendors are offering managed Knative services; see the [Knative offerings](../knative-offerings.md) page for a full list.
```

   {{< tabs name="serving_networking" default="Istio" >}}
{{% tab name="Ambassador" %}}

The following commands install Ambassador and enable its Knative integration:

1. Create a namespace to install Ambassador in:

   ```bash
   kubectl create namespace ambassador
   ```

2. Install Ambassador:

   ```bash
   kubectl apply --namespace ambassador \
     --filename https://getambassador.io/yaml/ambassador/ambassador-rbac.yaml \
     --filename https://getambassador.io/yaml/ambassador/ambassador-service.yaml
   ```

3. Give Ambassador the required permissions:

   ```bash
   kubectl patch clusterrolebinding ambassador -p '{"subjects":[{"kind": "ServiceAccount", "name": "ambassador", "namespace": "ambassador"}]}'
   ```

4. Enable Knative support in Ambassador:

   ```bash
   kubectl set env --namespace ambassador  deployments/ambassador AMBASSADOR_KNATIVE_SUPPORT=true
   ```

5. To configure Knative Serving to use Ambassador by default:

    ```bash
    kubectl patch configmap/config-network \
      --namespace knative-serving \
      --type merge \
      --patch '{"data":{"ingress.class":"ambassador.ingress.networking.knative.dev"}}'
    ```

> 6. Fetch the External IP or CNAME

Reviewer: ```suggestion
6. Fetch the External IP or CNAME:
```

Author: done everywhere

   ```bash
   kubectl --namespace ambassador get service ambassador
   ```

   Save this for configuring DNS below.

{{< /tab >}}

{{% tab name="Contour" %}}

> The following commands install Contour and enables its Knative integration:

Reviewer: ```suggestion
The following commands install Contour and enables its Knative integration.
```

1. Install the Knative Contour controller (includes Contour):

   ```bash
   kubectl apply --filename https://raw.githubusercontent.com/knative/serving/{{< version >}}/third_party/contour-latest/contour.yaml
   ```

2. To configure Knative Serving to use Contour by default:

    ```bash
    kubectl patch configmap/config-network \
      --namespace knative-serving \
      --type merge \
      --patch '{"data":{"ingress.class":"contour.ingress.networking.knative.dev"}}'
    ```

> 3. Fetch the External IP or CNAME

Reviewer: ```suggestion
3. Fetch the External IP or CNAME:
```

   ```bash
   kubectl --namespace projectcontour get service envoy-external
   ```

   Save this for configuring DNS below.

{{< /tab >}}

{{% tab name="Gloo" %}}

> _For a detailed guide on Gloo integration, see [here](https://docs.solo.io/gloo/latest/installation/knative/)_

Reviewer: ```suggestion
_For a detailed guide on Gloo integration, see [Installing Gloo for Knative](https://docs.solo.io/gloo/latest/installation/knative/) in the Gloo documentation._
```

> The following commands install Gloo and enables its Knative integration:

Reviewer: ```suggestion
The following commands install Gloo and enable its Knative integration.
```

Author: done everywhere

1. Make sure `glooctl` is installed (version 1.3.x and higher recommended):

   ```bash
   glooctl version
   ```

   If it is not installed, follow the instructions [here](https://docs.solo.io/gloo/latest/installation/knative/#install-command-line-tool-cli)

2. Install Gloo and the Knative integration:

    ```bash
    glooctl install knative --install-knative=false
    ```

> 3. Fetch the External IP or CNAME

Reviewer: ```suggestion
3. Fetch the External IP or CNAME:
```

   ```bash
   kubectl get svc -ngloo-system knative-external-proxy
   ```

   Save this for configuring DNS below.

{{< /tab >}}

{{% tab name="Istio" %}}
The following commands install Istio and enables its Knative integration:

<!-- TODO(https://github.com/knative/docs/issues/2166): Create streamlined instructions to inline -->

1. [Installing Istio for Knative](./installing-istio.md)

2. Install the Knative Istio controller:

   ```bash
   kubectl apply --filename https://github.com/knative/serving/releases/download/{{< version >}}/serving-istio.yaml
   ```

> 3. Fetch the External IP or CNAME

Reviewer: ```suggestion
3. Fetch the External IP or CNAME:
```

   ```bash
   kubectl --namespace istio-system get service istio-ingressgateway
   ```

   Save this for configuring DNS below.

{{< /tab >}}

{{% tab name="Kourier" %}}
The following commands install Kourier and enables its Knative integration:

1. Install the Knative Kourier controller:

   ```bash
   kubectl apply --filename https://raw.githubusercontent.com/knative/serving/{{< version >}}/third_party/kourier-latest/kourier.yaml
   ```

2. To configure Knative Serving to use Kourier by default:

    ```bash
    kubectl patch configmap/config-network \
      --namespace knative-serving \
      --type merge \
      --patch '{"data":{"ingress.class":"kourier.ingress.networking.knative.dev"}}'
    ```

> 3. Fetch the External IP or CNAME

Reviewer: ```suggestion
3. Fetch the External IP or CNAME:
```

   ```bash
   kubectl --namespace kourier-system get service kourier
   ```

   Save this for configuring DNS below.

{{< /tab >}}
{{< /tabs >}}

1. Configure DNS

   <!-- This indentation is important for things to render properly. -->
   {{< tabs name="serving_dns" >}}
{{% tab name="Magic DNS (xip.io)" %}}
We ship a simple Kubernetes Job called "default domain" that will (see caveats) configure Knative Serving to use <a href="http://xip.io">xip.io</a> as the default DNS suffix.

```bash
kubectl apply --filename https://github.com/knative/serving/releases/download/{{< version >}}/serving-default-domain.yaml
```

**Caveat**: This will only work if the cluster LoadBalancer service exposes an IPv4 address, so it will not work with IPv6 clusters, AWS, or local setups like Minikube.  For these, see "Real DNS".
{{< /tab >}}

{{% tab name="Real DNS" %}}
To configure DNS for Knative, take the External IP or CNAME from setting up networking, and configure it with your DNS provider as follows:

> - If the networking layer produced an External IP address, then configure a wildcard A record for the domain:

Reviewer: Is "a wildcard A record" correct, or should it be "a wildcard record"?

Author: I made it `A` to clarify.

    ```
    # Here knative.example.com is the domain suffix for your cluster
    *.knative.example.com == A 35.233.41.212
    ```

- If the networking layer produced a CNAME, then configure a CNAME record for the domain:

    ```
    # Here knative.example.com is the domain suffix for your cluster
    *.knative.example.com == CNAME a317a278525d111e89f272a164fd35fb-1510370581.eu-central-1.elb.amazonaws.com
    ```

> Once your DNS provider has been configured, direct Knative to use that domain with:

Reviewer: ```suggestion
Once your DNS provider has been configured, direct Knative to use that domain:
```

```bash
# Replace knative.example.com with your domain suffix
kubectl patch configmap/config-domain \
  --namespace knative-serving \
  --type merge \
  --patch '{"data":{"knative.example.com":""}}'
```

{{< /tab >}}
{{< /tabs >}}

1. Monitor the Knative components until all of the components show a `STATUS` of `Running` or `Completed`:

   ```bash
   kubectl get pods --namespace knative-serving
   ```

At this point, you have a basic installation of Knative Serving!

### Optional Serving extensions

{{< tabs name="serving_extensions" >}}
{{% tab name="HPA autoscaling" %}}

Knative also supports the use of the Kubernetes Horizontal Pod Autoscaler (HPA)
for driving autoscaling decisions.  The following command will install the
components needed to support HPA-class autoscaling:

```bash
kubectl apply --filename https://github.com/knative/serving/releases/download/{{< version >}}/serving-hpa.yaml
```

<!-- TODO(https://github.com/knative/docs/issues/2152): Link to a more in-depth guide on HPA-class autoscaling -->

{{< /tab >}}

{{% tab name="TLS with cert-manager" %}}

> Knative supports automatically provisioning TLS certificates via cert-manager.  The following commands will install the components needed to support the provisioning of TLS certificates via cert-manager:

Reviewer: ```suggestion
Knative supports automatically provisioning TLS certificates via [cert-manager](https://cert-manager.io/docs/).  The following commands will install the components needed to support the provisioning of TLS certificates via cert-manager:
```

Reviewer: Please check to make sure that's the right cert-manager :)

Author: it is!
